### PR TITLE
Restructure code and add some tests.

### PR DESCRIPTION
**Pull request by mpoeter**

### CODE REVIEW

File: include/mellon/completion-queue.h [hunk 1]
```diff
@@ -15,36 +15,35 @@ struct completion_context : std::enable_shared_from_this<completion_context<Fut>
   using value_type = typename trait::value_type;
 
   void register_future(Fut&& f) {
-    pending_futures.fetch_add(1, std::memory_order_release);
+  // mpoeter - should we change this signature so that it can also take a future-temporary?
```
Comment (maierlars) [inline include/mellon/completion-queue.h]: Yes!

File: include/mellon/impl/extensions.h [hunk 2]
```diff
@@ -52,6 +53,7 @@ struct FUTURES_EMPTY_BASE future_type_based_extensions<expect::expected<T>, Fut,
    * @param f
    * @return
    */
+  // mpoeter - do we need this code or should we remove it?
```
Comment (maierlars) [inline include/mellon/impl/extensions.h]: Please remove.

File: include/mellon/impl/extensions.h [hunk 3]
```diff
@@ -23,6 +23,7 @@ struct FUTURES_EMPTY_BASE future_type_based_extensions<expect::expected<T>, Fut,
             typename R = std::invoke_result_t<F, U&&>>
   auto then(F&& f) && noexcept {
     // TODO what if `F` returns an `expected<U>`. Do we want to flatten automagically?
+    // mpoeter - I think it would make sense!
```
Comment (maierlars) [inline include/mellon/impl/extensions.h:26]: I dislike doing things automagically. We should add an extra function for that.

File: include/mellon/impl/extensions.h [hunk 4]
```diff
@@ -60,6 +62,7 @@ struct FUTURES_EMPTY_BASE future_type_based_extensions<expect::expected<T>, Fut,
     return std::move(self()).and_capture(std::forward<F>(f));
   }*/
 
+  // mpoeter - documentation missing
```
Comment (maierlars) [inline include/mellon/impl/extensions.h:48]: Would you like to provide it? Otherwise I will do...

File: include/mellon/impl/extensions.h (hunk 3 above)
Comment (mpoeter) [reply include/mellon/impl/extensions.h:26]: Alright, then I'll remove my comment and the TODO.

File: include/mellon/impl/extensions.h (hunk 2 above)
Comment (mpoeter) [reply include/mellon/impl/extensions.h]: Done.

File: include/mellon/completion-queue.h (hunk 1 above)
Comment (mpoeter) [reply include/mellon/completion-queue.h]: Done.

File: include/mellon/impl/extensions.h (hunk 4 above)
Comment (mpoeter) [reply include/mellon/impl/extensions.h:48]: I thought about that, but I am not 100% sure what these functions do/should do, so I would prefer if you could do it.

**Pull request merged**